ttcn3/v2/syntax/syntax.go

```diff
@@ -297,6 +303,8 @@ func (t *tree) position(pos int) Position {
 func (t *tree) searchLines(pos int) int {
 	// TODO(5nord) add line cache
 	i, j := 0, len(t.lines)
+	t.cachedLineBegin = i
+	t.cachedLineEnd = j
```

Comment:
I think those two lines are the reason why the benchmarks are not as good as expected and why the tests fail:
`i` and `j` represent _lines_, while `cachedLineBegin` and `cachedLineEnd` are supposed to represent (file) _offsets_. In line 288 you're essentially comparing lines with offsets:
```
if t.cachedLineBegin <= pos && pos < t.cachedLineEnd
```
I think those two variables should be set after you found the proper line:
 ```suggestion
```

Reply:
i have now put the comparison steps in searchLines function as below so searchLines function will return cacheLine
```
	if t.cachedLineBegin <= i && j < t.cachedLineEnd {
		return t.cacheLine
	}
```
is this correct ?

Reply:
Almost. Instead of `i` and `j` you should compare with `pos`:
```go
if t.cachedLineBegin <= pos && pos < t.cachedLineEnd {
        return t.cacheLine
```

I'd like to explain a bit how location handling works, this will clear things up:

### Overview
One core functionality of this packages is to provide start and end position for syntax nodes. We call this a text span and it includes a file name, line numbers, column numbers and the file offsets. We _could_ store this information inside the node object. For example:
```go
type Node struct {
        // ...
        Span struct {
                Filename string
                Begin, End struct {
                        Line, Column, Offset int
                }
        }
}
```

This approach is strait-forward and easy, but also quite wasteful. It would add 64 bytes of repetitive data to every node and sometimes we need this detailed information only for a few nodes.

### So, how can we improve this situation?
Instead of storing the whole span, we store the begin and end offsets (8 bytes instead of 64) and provide an algorithm to compute the detailed information only when needed. For example the node would look like this:
```go
type Node struct {
        // ...
        Begin, End int32
}
```

### How to we get from offset to line numbers?

When creating the syntax tree, we keep track of newline characters (`\n`) and store the start-offset for every line in a slice.

**Example: Lines slice**

The input `Foo\nBar Baz\nFoobar\n` contains three lines, the first line starts is at offset 0, the second line starts at offset 4, the last line starts at offset 12. The resulting lines slice would like this:
```go
lines := []int{0, 4, 12}
```

**Example: Lookup**

In which line is the word `Baz`? The know from the syntax nodes, that the word begins at offset 8.
Now we search in the lines slice:
```go
for i := range lines {
        if 8 < lines[i] {
                return i-1
        }
}  
return len(lines)-1
```

We need to subtract 1 from the found index, because this loop exits at one line after the one we were actually searching for; in this case it breaks at the third line (`i==2`).

### Binary Search

Our files have 300-1500 lines on average. As a consequence the described lookup becomes very inefficient, because it's just a linear search. We can do _much_ better if we use a binary search algorithm instead.
Here are some articles I found explaining binary search:
* [Everything you need to know](https://towardsdatascience.com/everything-you-need-to-know-about-the-binary-search-algorithm-6bc4f9a3127d)
 * [Geeks For Geeks](https://www.geeksforgeeks.org/binary-search/)
 
The `tree.searchLines` implements  such a binary search. The variables `i` and `j` are the lower- and the higher limits required for the binary search. That's why `i` starts at 0 and walks towards the end of the lines slice, while `j` starts end the end (`len(t.lines)`) and walks towards the beginning of the lines lines.


### Caching

The caching is independent of those binary search variables. The meaning of those is:
* `cachedLineBegin`: The offset at which the cached line starts.
* `cachedLineEnd`: The offset at which the cached line ends.
* `cachedLine`: The cached line number.

Reply:
thanks for detailed explanation.